homeassistant/components/efergy/sensor.py

```diff
@@ -96,68 +136,55 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 class EfergySensor(SensorEntity):
     """Implementation of an Efergy sensor."""
 
-    def __init__(self, sensor_type, app_token, utc_offset, period, currency, sid=None):
+    def __init__(
+        self,
+        app_token: Any,
+        utc_offset: str,
+        period: str,
+        currency: str,
+        description: SensorEntityDescription,
+        sid: str = None,
+    ) -> None:
         """Initialize the sensor."""
+        self.entity_description = description
         self.sid = sid
         if sid:
-            self._name = f"efergy_{sid}"
-        else:
-            self._name = SENSOR_TYPES[sensor_type][0]
-        self.type = sensor_type
+            self._attr_name = f"efergy_{sid}"
         self.app_token = app_token
         self.utc_offset = utc_offset
-        self._state = None
         self.period = period
-        self.currency = currency
-        if self.type == "cost":
-            self._unit_of_measurement = f"{self.currency}/{self.period}"
-        else:
-            self._unit_of_measurement = SENSOR_TYPES[sensor_type][1]
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        return self._state
-
-    @property
-    def native_unit_of_measurement(self):
-        """Return the unit of measurement of this entity, if any."""
-        return self._unit_of_measurement
-
-    def update(self):
+        if description.key == CONF_COST:
+            self._attr_native_unit_of_measurement = f"{currency}/{period}"
+
+    def update(self) -> None:
         """Get the Efergy monitor data from the web service."""
         try:
-            if self.type == "instant_readings":
+            if self.entity_description.key == CONF_INSTANT:
                 url_string = f"{_RESOURCE}getInstant?token={self.app_token}"
```

Comment:
Existing for a future PR:

- `url_string` should be abstracted into the pypi package.  Home Assistant should not know about these implementation details
- Only the code that can throw should be in the try block